patroni/postgresql.py

```diff
@@ -1645,52 +1717,77 @@ def basebackup(self, conn_url, env):
 
         return ret
 
-    def pick_synchronous_standby(self, cluster):
-        """Finds the best candidate to be the synchronous standby.
-
-        Current synchronous standby is always preferred, unless it has disconnected or does not want to be a
-        synchronous standby any longer.
-
-        :returns tuple of candidate name or None, and bool showing if the member is the active synchronous standby.
+    def current_sync_state(self, cluster):
+        """Returns current synchronous replication state as a dict:
+        * numsync: number of synchronous nodes requested(including leader)
+        * sync: set of nodes potentially being synced to
+        * active: set of nodes that are sync capable
         """
-        current = cluster.sync.sync_standby
-        current = current.lower() if current else current
+        sync_standby_names = self.query("SHOW synchronous_standby_names")
+        result = parse_sync_standby_names(sync_standby_names)
```

Comment:
Why do we need to parse it all the time? Is it some kind of protection from unexpected changes from outside?

Reply:
Basically yes. It helps us automatically resolve the situation if due to a bug or administrator intervention our understanding of the state goes out of sync. If we aren't too worried about that we could cache the state and avoid spamming the database with unnecessary queries. What do you think?

Reply:
The problem is that it could be changed with 'ALTER SYSTEM' which takes precedence over postgresql.conf :(

Reply:
We could detect that from pg_settings. Now the question is what should the system do in that situation? Override the administrator with `ALTER SYSTEM RESET synchronous_standby_names`, make a quorum state to match the synchronization state, just complain in the logs where nobody is looking anyway, panic with `sys.exit(1)`?

Reply:
Looks like executing `ALTER SYSTEM RESET synchronous_standby_names` is the best we can do.
In addition to that we can write warning into the log with previous value.